build.sh

```diff
@@ -5,80 +5,109 @@
 if [ "$EUID" -ne 0 ]; then
     echo "Please run this script as root user"
     exit
 fi
 
+# 'set -eo pipefail' means we need to be careful with grep and zero search results
+# because grep exits with an error code, causing the pipeline to fail, script exit
+#
+# a workaround is to enclose grep as follows, where e is search term:
+#
+# cmd | { grep e || :; } | cmd
+#
+set -eo pipefail
+
+exit_error() {
+    echo "$*" 1>&2
+    exit 1;
+}
+
+usage() {
+    cat <<-EOF
+	Usage: $(basename "${BASH_SOURCE[0]}") [-hbu] [-k /path/to/authorized_keys] 
+	
+	-h Show help
+	-b Build without uploading
+	-u Build with upload to remote host
+	-k /path/to/authorized_keys (can safely ignore, another opportunity to copy in SSH keys on image boot!)
+	
+	EOF
+}
+
 # we must be on freebsd
 what_os_am_i="$(uname)"
 if [ "${what_os_am_i}" != "FreeBSD" ]; then
-    echo "Please run on FreeBSD only"
-    exit
+    exit_error "Please run on FreeBSD only"
 fi
 
 if [ $# -lt 1 ]; then
-  1>&2 echo "
-Usage: $0 -i access-ip -f /path/to/authorized_keys -u [0/1]
-
-       Example with upload on:
-       $0 -i \"10.0.0.1\" -f /root/copy_in_auth_keys -u 1
-"
-  exit 1
+  1>&2 exit_error "$(usage)"
 fi
 
 # get command line flags
-while getopts i:f:u: flag
+while getopts hbuk: flag
 do
     case "${flag}" in
-        i) MYACCESSIP="${OPTARG}";;
-        f) AUTHKEYFILE="${OPTARG}";;
-        u) UPLOAD="${OPTARG}";;
+        h) 
+           usage
+           exit 0
+           ;;
+        b)
+           UPLOAD="NO"
+           ;;
+        u) 
+           UPLOAD="YES"
+           ;;
+        k) 
+           AUTHKEYFILE="${OPTARG}"
+           ;;
+        *) 
+           exit_error "$(usage)"
+           ;;
     esac
 done
+shift "$((OPTIND-1))"
 
-# VARS
-if [ -z "${MYACCESSIP}" ]; then
-    MYACCESSIP="*"
-fi
+# Set default values if not set
+# XXX: this might be unnecessary
 if [ -z "${AUTHKEYFILE}" ]; then
     touch authorized_keys_in
     AUTHKEYFILE="authorized_keys_in"
 fi
-if [ -z "${UPLOAD}" ]; then
-    UPLOAD=0
-fi
 
-# VARS
+# General Variables
 BASEDIR="$PWD"
 CDMOUNT="cd-rom"
-CHECKMOUNTCD1=$(mount | grep cd-rom | awk -F" " '{print $1}')
+CHECKMOUNTCD1="$(mount | { grep "${CDMOUNT}" || :; } | awk '{print $1}')"
 FREEBSDISOSRC="https://download.freebsd.org/releases/amd64/amd64/ISO-IMAGES/13.1/FreeBSD-13.1-RELEASE-amd64-dvd1.iso"
 FREEBSDISOFILE="FreeBSD-13.1-RELEASE-amd64-dvd1.iso"
 MFSBSDDIR="mfsbsd"
 MYRELEASE="13.1-RELEASE"
 MYARCH="amd64"
-OUTIMG="mfsbsd-${MYRELEASE}-${MYARCH}.img"
-OUTISO="mfsbsd-${MYRELEASE}-${MYARCH}.iso"
+OUTIMG="mfsbsd-${MYRELEASE}-${MYARCH}.img"    # not in use
+OUTISO="mfsbsd-${MYRELEASE}-${MYARCH}.iso"    # in use
 MYBASE="${BASEDIR}/${CDMOUNT}/usr/freebsd-dist"
-MYREMOTECONFIG="${BASEDIR}/settings.cfg"
 MYCUSTOMDIR="${BASEDIR}/customfiles"
 
 # make sure we're in base directory
 cd "${BASEDIR}" || exit
 
 # check remote settings
-if [ -f "${MYREMOTECONFIG}" ]; then
-    source "${MYREMOTECONFIG}"
+# shellcheck source=/dev/null
```

Comment:
## General

I would suggest to move the shellcheck comment directly above the source command, so it's absolutely clear where it refers to.

## My first suggestion (read only if curious)

I would also just read the parameters in when UPLOAD is YES (right now, that's the only time when it's needed, even though this might change in the future) and validate parameters.

Finally, config variables should be set to empty to make sure they're set using settings.sh (otherwise there could be an empty settings file and variables come from the environment, which could be confusing):

```suggestion
if [ "$UPLOAD" = "YES" ]; then
    CFG_SSH_REMOTEHOST=
    CFG_SSH_REMOTEPATH=
    if [ -f "${BASEDIR}/settings.sh" ]; then
        # shellcheck source=/dev/null
        source "${BASEDIR}/settings.sh"
    else
        exit_error "Please copy settings.sh.sample to settings.sh and set parameters"
    fi
    if [ -z "${CFG_SSH_REMOTEHOST}" ]; then
        exit_error "CFG_SSH_REMOTEHOST not configured in settings.sh"
    fi
fi
```

## My preferred solution

The alternative would be to not to wipe CFG_SSH_... at all, but in this case I would suggest to make settings.sh completely optional (as users could pass the variables in over the command line without creating settings.sh).

If you want to go that route, the suggestion would be:

```suggestion
if [ -f "${BASEDIR}/settings.sh" ]; then
    # shellcheck source=/dev/null
    source "${BASEDIR}/settings.sh"
fi
if [ "$UPLOAD" = "YES" ] && [ -z "${CFG_SSH_REMOTEHOST}" ]; then
    exit_error "CFG_SSH_REMOTEHOST not configured in settings.sh"
fi
```

To be honest, I like this second option better, as it allows users to do:

    CFG_SSH_REMOTEHOST=myhost ./build.sh -u

Reply:
I just had another idea: What about providing a default for CFG_SSH_REMOTEHOST? (e.g., `depenguin-me`).

Then all users would need to do is add an entry to their `~/.ssh/config` like this one:

```
Host depenguin-me
   User myuser
   HostName myactualhost.example.org
   IdentityFile ~/.ssh/byprivatekey
```

And they could run `./build.sh -u`

In this case, the code would be:

```suggestion
if [ -f "${BASEDIR}/settings.sh" ]; then
    # shellcheck source=/dev/null
    source "${BASEDIR}/settings.sh"
fi
if [ -z "${CFG_SSH_REMOTEHOST}" ]; then
    CFG_SSH_REMOTEHOST=depenguin-me
fi
```

**Note:** Suggestions don't really work when there are deleted lines, so you can't simply apply it in github's web u

Reply:
adopting this approach. 

tests:
- build image works
- run image with flags for keys works (login, sudo to root)

Reply:
thanks for the feedback!